# SQL

**Цель:** 
Вывести ознакомительную информацию из базы данных для дальнейшего формулирования ценностного предложения для нового продукта.

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

[Содержание:](#content)
<a id='content'></a>

[Описание данных](#sql) 

[Загрузка данных](#sql1)
    
[Исследование таблицы — выведение первых строк](#sql2)

[Посчитаем, сколько книг вышло после 1 января 2000 года](#sql3)

[Для каждой книги посчитаем количество обзоров и среднюю оценку](#sql4)

[Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры](#sql5)

[Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками.](#sql6)

[Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок](#sql7)

[Выводы](#sql8)

### Описание данных
[](#sql)
<a id='sql'></a>


**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

- Посчитаем, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитаем количество обзоров и среднюю оценку;
- Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками;
- Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

[К содержанию](#content)

### Загрузка данных
[](#sql1)
<a id='sql1'></a>

Импортируем библиотеки

In [1]:
import pandas as pd
from sqlalchemy import create_engine 

Устанавливаем параметры

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных


In [3]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db']) 

Сохраняем коннектор

In [4]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Чтобы выполнить SQL-запрос, используем Pandas:

pd.io.sql.read_sql(query, con = engine)

[К содержанию](#content)

### Исследование таблицы — выведение первых строк
[](#sql2)
<a id='sql2'></a>

Распечатаем первые 5 строк таблицы books:

In [5]:
query_1 = '''
         SELECT *
         FROM books
         LIMIT 5
        '''
books = pd.io.sql.read_sql(query_1, con = engine)
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Распечатаем первые 5 строк таблицы authors:

In [6]:
query_2 = '''
         SELECT *
         FROM authors
         LIMIT 5
        '''
authors = pd.io.sql.read_sql(query_2, con = engine)
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Распечатаем первые 5 строк таблицы publishers:

In [7]:
query_3 = '''
         SELECT *
         FROM publishers
         LIMIT 5
        '''
publishers = pd.io.sql.read_sql(query_3, con = engine)
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Распечатаем первые 5 строк таблицы ratings:

In [8]:
query_4 = '''
         SELECT *
         FROM ratings
         LIMIT 5
        '''
ratings = pd.io.sql.read_sql(query_4, con = engine)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Распечатаем первые 5 строк таблицы reviews:

In [9]:
query_5 = '''
         SELECT *
         FROM reviews
         LIMIT 5
        '''
reviews = pd.io.sql.read_sql(query_5, con = engine)
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

При выведении первых строк таблиц, отметим, что названия столбцов адекватно отображают своё содержимое. Явных признаков некорректного заполнения не обнаружено.

[К содержанию](#content)

### Посчитаем, сколько книг вышло после 1 января 2000 года
[](#sql3)
<a id='sql3'></a>

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [10]:
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Посчитаем, сколько книг вышло после 1 января 2000 года;

In [11]:
query_6 = '''
         SELECT 
               COUNT(book_id)
         FROM  books
         WHERE 
               publication_date >= '01-01-2000'
        '''
books_after_2000 = pd.io.sql.read_sql(query_6, con = engine)
books_after_2000

count
0    821

По данным таблицы books, после первого января 2000 года вышлa 821 книгa.

[К содержанию](#content)

### Для каждой книги посчитаем количество обзоров и среднюю оценку
[](#sql4)
<a id='sql4'></a>

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [12]:
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [14]:
query_7 = '''
         SELECT
              SQB.book_id,
              SQB.title,
              AVG(SQB2.avg_rating) AS rating_avg,
              COUNT(DISTINCT SQB3.count_review) AS count_review
         FROM 
              (
              SELECT
                       books.book_id,
                       books.title
              FROM  
                       books   
              GROUP BY
                       books.book_id
                       
               ORDER BY
                      books.title
                ) AS SQB
               
                       LEFT JOIN (
                      
                                    SELECT
                                           ratings.book_id,
                                           AVG(ratings.rating) AS avg_rating
                                    FROM 
                                           ratings
                                    GROUP BY  
                                           ratings.book_id
                                    ORDER BY 
                                           avg_rating DESC  
                                    
                                  
                                    )  AS SQB2 ON SQB2.book_id = SQB.book_id
                                    
                                                    INNER JOIN (
                                                                   SELECT
                                                                         reviews.book_id,
                                                                         reviews.review_id AS count_review      
                                                                   FROM 
                                                                         reviews
                                                                   GROUP BY 
                                                                         reviews.book_id,
                                                                         count_review 
                                                                   ORDER BY 
                                                                         count_review DESC 
                                                                   
                                                                    ) AS SQB3 ON SQB3.book_id = SQB2.book_id
               
               
                                                             
        GROUP BY                        
                SQB.book_id,
                title
            
                
                
        ORDER BY
                count_review DESC,
                rating_avg DESC
                
                
        '''
reviews_and_average_rating = pd.io.sql.read_sql(query_7, con = engine)
reviews_and_average_rating

book_id                                              title  rating_avg  \
0        948                            Twilight (Twilight  #1)    3.662500   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634   
2        299  Harry Potter and the Chamber of Secrets (Harry...    4.287500   
3        656                                     The Book Thief    4.264151   
4        734                                   The Glass Castle    4.206897   
..       ...                                                ...         ...   
989      186                     Death: The High Cost of Living    3.000000   
990      446                                Moo  Baa  La La La!    3.000000   
991        2                 1 000 Places to See Before You Die    2.500000   
992      318                                  History of Beauty    2.500000   
993      943               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1    2.500000   

     count_review  
0               7  
1               6  
2               6  
3               6  
4               6  
..            ...  
989             1  
990             1  
991             1  
992             1  
993             1  

[994 rows x 4 columns]

Для каждой книги посчитали количество обзоров и среднюю оценку

Интересно было бы узнать, какие книги являются топовыми с точки зрения максимума обзоров и наибольшего показателя средней оценки. Топ-5 книг (первые пять книг таблицы books).

Топ-5 книг (на самом деле первые пять книг таблицы books) с наибольшими показателями средней оценки, и максимумом обзоров.

In [15]:
query_8 = '''
         SELECT
               books.book_id,
               books.title AS name_book,
               COUNT(DISTINCT reviews.review_id) AS count_review,
               AVG(ratings.rating) AS avg_rating
         FROM  books
               LEFT JOIN reviews ON reviews.book_id = books.book_id
               LEFT JOIN ratings ON ratings.book_id = books.book_id      
         GROUP BY
               books.book_id
         ORDER BY
               avg_rating DESC,
               count_review DESC
         LIMIT 5      
               
        '''
top_5_rating_review = pd.io.sql.read_sql(query_8, con = engine)
top_5_rating_review

book_id                                     name_book  count_review  \
0       17                 A Dirty Job (Grim Reaper  #1)             4   
1      444  Moneyball: The Art of Winning an Unfair Game             3   
2      553       School's Out—Forever (Maximum Ride  #2)             3   
3       57               Act of Treason (Mitch Rapp  #9)             2   
4       20         A Fistful of Charms (The Hollows  #4)             2   

   avg_rating  
0         5.0  
1         5.0  
2         5.0  
3         5.0  
4         5.0

- A Dirty Job (Grim Reaper #1) - Кристофер Мур "Грязная работа"
- Moneyball: The Art of Winning an Unfair Game - Майкл Льюис "Денежный шар"
- School's Out—Forever (Maximum Ride #2) - Джеймс Паттерсон "Максимальная поездка: Школа закрыта Навсегда"
- Act of Treason (Mitch Rapp #9) - Винс Флинн "Акт государственной измены"
- A Fistful of Charms (The Hollows #4) - Ким Харрисон "Пригоршня прелестей"	

<img src=https://kbimages1-a.akamaihd.net/af30029e-070d-4f4d-8d7c-d00b65459176/1200/1200/False/a-dirty-job-2.jpg>

**Грязная работа**

- Автор:	Кристофер Мур
- Художник обложки:	Уильям Стейл
- Страна:	США
- Язык:	Английский
- Жанр:	Юмор, комическая фантазия, детективная фантастика, приключенческая фантастика, абсурдистская фантастика, ужасы, городская фантазия

"Грязная работа" - роман американского писателя Кристофера Мура, опубликованный в 2006 году. Отражая абсурдистские тенденции автора, содержание романа в значительной степени опирается на его собственный опыт в удовлетворении потребностей близких родственников и друзей на этапах умирания.

<img src=https://www.incomediary.com/wp-content/uploads/2012/01/moneyball.jpg>

**Денежный шар**

- Автор:	Майкл Льюис
- Страна:	США
- Язык:	Английский
- Жанр:	Нехудожественная литература

"Moneyball: The Art of Winning an Unfair Game"-это книга Майкла Льюиса, опубликованная в 2003 году, о бейсбольной команде Oakland Athletics и ее генеральном менеджере Билли Бине. Его фокус-аналитический, основанный на фактических данных, саберметрический подход команды к созданию конкурентоспособной бейсбольной команды, несмотря на небольшой бюджет Окленда. Фильм по книге Льюиса с Брэдом Питтом и Джоной Хилл в главных ролях вышел на экраны в 2011 году.

<img src=https://i.pinimg.com/originals/93/03/1c/93031cdc2352bea69da324f39877ed59.jpg>

**Максимальная поездка: Школа закрыта Навсегда**

- Автор:	Джеймс Паттерсон
- Художник обложки:	Дэвид Перри, Пит Тернер и Роб Мэтисон
- Страна:	США
- Язык:	Английский
- Жанр:	Молодые взрослые, научная фантастика

"Maximum Ride: School's Out—Forever"-вторая книга Джеймса Паттерсона из научно-фантастической приключенческой серии "Максимальная поездка", опубликованная издательством Литтл Браун. Книга была выпущена в США и Великобритании в мае 23, 2006.[2][3] Сериал рассказывает о "Стае", группе гибридов человека и птицы, скрывающихся от ученых, которые их создали. "Стая" состоит из 6 детей, обладающих различными способностями; Ангел может читать мысли, Надж может рассказать больше информации о любой вещи, если она может прикоснуться к ней, а Игги, пока он слеп, все еще может создавать самые сложные бомбы.

<img src=https://offtheshelf.com/app/uploads/2016/10/Act-of-Treason.jpg>

**Акт государственной измены**

- Автор: Винс Флинн
- Страна: США
- Язык: Английский
- Жанр:	Политический триллер

Митч Рэпп-вымышленный тайный агент ЦРУ по борьбе с терроризмом. Основное внимание Рэппа сосредоточено на предотвращении террористических нападений на Соединенные Штаты, и он представлен как агрессивный оперативник, готовый принять меры, которые являются более крайними, чем могут быть сочтены общепризнанными. Его постоянное недовольство процедурами и бюрократией является главной темой на протяжении всей серии.

<img src=https://c.booko.info/covers/3eeedefeb37ceca8/v/600.jpeg>

**Пригоршня прелестей**

- Автор: Ким Харрисон
- Страна: США
- Язык: Английский
- Жанр: Городское фэнтези

Злые ночные твари, которые бродят по Цинциннати, презирают ведьму и охотницу за головами Рэйчел Морган. Ее новая репутация в темных искусствах сводится к тому, что люди и нежить одинаково склоняют головы с намерением овладеть, уложить в постель и убить ее-не обязательно в таком порядке.

Теперь смертный любовник, бросивший Рейчел, вернулся, преследуемый своим тайным прошлым. И есть те, кто жаждет того, чем владеет Ник, - дикие звери, готовые при необходимости уничтожить Лощины и всех, кто в них находится.

Вынужденная держаться в тени или вечно страдать от гнева мстительного демона, Рейчел, тем не менее, должна действовать быстро. Ибо стая собирается впервые за тысячелетия, чтобы опустошать и править. И вдруг на карту поставлено нечто большее, чем душа Рейчел.



---

Занятная подборка. Для более пристального изучения этой тематики, видимо, не помешало бы узнать как и кем происходит присваивание рейтинга и в каких странах (стране?) происходит подсчёт обзоров. Посмотрим также без деталей на Топ-5 (или первых выведенных по наибольшим показателям по таблице) отдельно по максимальному количеству обзоров и самому высокому показателю среднего рейтинга. 

Обзоры:

In [16]:
query_9 = '''
         SELECT
               books.book_id,
               books.title AS name_book,
               COUNT(DISTINCT reviews.review_id) AS count_review,
               AVG(ratings.rating) AS avg_rating
         FROM  books
               LEFT JOIN reviews ON reviews.book_id = books.book_id
               LEFT JOIN ratings ON ratings.book_id = books.book_id      
         GROUP BY
               books.book_id
         ORDER BY
               count_review DESC
         LIMIT 5      
               
        '''
top_5_review = pd.io.sql.read_sql(query_9, con = engine)
top_5_review

book_id                                          name_book  count_review  \
0      948                            Twilight (Twilight  #1)             7   
1      497                          Outlander (Outlander  #1)             6   
2      207                                    Eat  Pray  Love             6   
3      299  Harry Potter and the Chamber of Secrets (Harry...             6   
4      302  Harry Potter and the Prisoner of Azkaban (Harr...             6   

   avg_rating  
0    3.662500  
1    4.125000  
2    3.395833  
3    4.287500  
4    4.414634

- Twilight (Twilight #1) - Стефани Майер "Сумерки";
- Outlander (Outlander #1) - Гэблдон Диана "Чужестранка";
- Eat Pray Love - Элизабет Гилберт "Есть, молиться, любить";
- Harry Potter and the Chamber of Secrets - Джоан Роулинг "Гарри Поттер и Тайная комната";
- Harry Potter and the Prisoner of Azkaban - Джоан Роулинг "Гарри Поттер и узник Азкабана".

Обратим внимание на то, что каждое из произведений, попавших в список первых строк по максимальному количеству обзоров, обязательно имеет экранизацию. Что покажет нам рейтинг?

Рейтинг:

In [17]:
query_10 = '''
         SELECT
               books.book_id,
               books.title AS name_book,
               COUNT(DISTINCT reviews.review_id) AS count_review,
               AVG(ratings.rating) AS avg_rating
         FROM  books
               LEFT JOIN reviews ON reviews.book_id = books.book_id
               LEFT JOIN ratings ON ratings.book_id = books.book_id      
         GROUP BY
               books.book_id
         ORDER BY
               avg_rating DESC
         LIMIT 5      
               
        '''
top_5_rating = pd.io.sql.read_sql(query_10, con = engine)
top_5_rating

book_id                                  name_book  count_review  \
0       55  A Woman of Substance (Emma Harte Saga #1)             2   
1       57            Act of Treason (Mitch Rapp  #9)             2   
2       17              A Dirty Job (Grim Reaper  #1)             4   
3       20      A Fistful of Charms (The Hollows  #4)             2   
4       62                              Alas  Babylon             2   

   avg_rating  
0         5.0  
1         5.0  
2         5.0  
3         5.0  
4         5.0

- A Woman of Substance (Emma Harte Saga #1)	- Барбара Брэдфорд "Состоятельная женщина";
- Act of Treason (Mitch Rapp #9) - Винс Флинн "Акт государственной измены"
- A Dirty Job (Grim Reaper #1) - Кристофер Мур "Грязная работа"
- A Fistful of Charms (The Hollows #4) - Ким Харрисон "Пригоршня прелестей"
- Alas Babylon	- Пэт Франк "Увы, Вавилон"

Обратим внимание на пересечение произведений по критерию самого высокого показателя среднего рейтинга и объединённого критерия - с наибольшими показателями средней оценки, и максимумом обзоров:
- Act of Treason (Mitch Rapp #9) - Винс Флинн "Акт государственной измены"
- A Dirty Job (Grim Reaper #1) - Кристофер Мур "Грязная работа"
- A Fistful of Charms (The Hollows #4) - Ким Харрисон "Пригоршня прелестей"

[К содержанию](#content)

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры
[](#sql5)
<a id='sql5'></a>

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;


In [18]:
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [19]:
query_11 = '''
         SELECT
               publishers.publisher_id,
               publishers.publisher AS publisher_name,
               COUNT(books.book_id) AS count_book
               
         FROM  books
               LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
         WHERE 
               books.num_pages > 50
         GROUP BY
               publisher_name,
               publishers.publisher_id
         ORDER BY 
               count_book DESC
         LIMIT 1;      
        '''
publishing_house= pd.io.sql.read_sql(query_11, con = engine)
publishing_house

publisher_id publisher_name  count_book
0           212  Penguin Books          42

<img src=https://i.pinimg.com/originals/34/de/89/34de89c42d2ec9b7cf1a8fb9aff8daa4.png>

В таблице publishers представлено максимальное количество изданий издательства Penguin Books - 42 произведения толще 50 страниц.

Penguin Books (рус. Пе́нгуин букс или Пингвин букс) — британское издательство, основанное в 1935 году в Лондоне сэром Алленом Лейном и его братьями Ричардом и Джоном. Главной заслугой издательства считают демократизацию книжного рынка, превратившую книгу из предмета роскоши в удовольствие, доступное массам.

[К содержанию](#content)

### Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками.
[](#sql6)
<a id='sql6'></a>

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [20]:
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [21]:
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [22]:
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [23]:
query_12 =  '''
          SELECT
              author,
              AVG(sub.avg_rating) AS high_avg_rating
          FROM
              (SELECT
                  author,
                  books.book_id,
                  AVG(rating) AS avg_rating
              FROM 
                  authors
              LEFT JOIN books ON books.author_id = authors.author_id
              LEFT JOIN ratings ON ratings.book_id = books.book_id
              GROUP BY
                  author,
                  books.book_id
              HAVING
                  COUNT(rating) > 50) AS sub
          GROUP BY
              author
          ORDER BY
              high_avg_rating DESC
        '''

author_with_the_highest_rating = pd.io.sql.read_sql(query_12, con = engine)
author_with_the_highest_rating

author  high_avg_rating
0                          J.K. Rowling/Mary GrandPré         4.283844
1                   Markus Zusak/Cao Xuân Việt Khương         4.264151
2                                      J.R.R. Tolkien         4.258446
3                                   Louisa May Alcott         4.192308
4                                        Rick Riordan         4.080645
5                                     William Golding         3.901408
6                                       J.D. Salinger         3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce         3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...         3.787879
9                                           Dan Brown         3.754540
10                                         Lois Lowry         3.750000
11            George Orwell/Boris Grabnar/Peter Škerl         3.729730
12                                    Stephenie Meyer         3.662500
13                                     John Steinbeck         3.622951

Автор с самой высокой средней оценкой книг (с 50 и более оценками) - Джоан Роулинг - кто бы сомневался. Посоперничать в популярности с ней мог бы, наверное, только Стивен Кинг.

<img src=https://businessman.ru/static/img/n/9/6/4/1/3/3/i/964133.jpg>

**Джоан Роулинг**

Джоан Роулинг CH, OBE, HonFRSE, FRCPE, FRSL (/ˈroʊlɪŋ/ ROH-ling;[1] родилась 31 июля 1965 года), более известная под псевдонимом Джоан Роулинг, - британская писательница, филантроп, кинопродюсер, телевизионный продюсер и сценарист. Она наиболее известна тем, что написала серию фэнтези о Гарри Поттере, которая получила множество наград и разошлась тиражом более 500 миллионов экземпляров[2][3], став самой продаваемой книжной серией в истории[4]. Книги легли в основу популярного киносериала, над которым Роулинг получила общее одобрение по сценариям[5] и была продюсером финальных фильмов[6]. Она также пишет детективы. под псевдонимом Роберт Гэлбрейт.

Посмотрим на фамилии других авторов с высокой средней оценкой книг (с учётом 50 или более оценок):

In [24]:
query_13 = '''
         SELECT
               authors.author_id,
               authors.author AS authors_name,
               AVG(ratings.rating) AS avg_rating

         FROM  books
               LEFT JOIN authors ON authors.author_id = books.author_id
               LEFT JOIN ratings ON ratings.book_id = books.book_id
         GROUP BY
               authors_name,
               authors.author_id
         HAVING
               COUNT(books.book_id) > 50
         ORDER BY 
               avg_rating DESC
        '''
authors_with_the_highest_rating = pd.io.sql.read_sql(query_13, con = engine)
authors_with_the_highest_rating

author_id                                       authors_name  avg_rating
0         236                         J.K. Rowling/Mary GrandPré    4.288462
1           3                                    Agatha Christie    4.283019
2         402                  Markus Zusak/Cao Xuân Việt Khương    4.264151
3         240                                     J.R.R. Tolkien    4.240964
4         499                           Roald Dahl/Quentin Blake    4.209677
5         376                                  Louisa May Alcott    4.203704
6         498                                       Rick Riordan    4.130952
7          39                                      Arthur Golden    4.107143
8         542                                       Stephen King    4.009434
9         302                                       John Grisham    3.971429
10        621                                    William Golding    3.901408
11        445                                    Nicholas Sparks    3.882883
12        291                                       Jodi Picoult    3.881579
13        536                                    Sophie Kinsella    3.877193
14        257                                    James Patterson    3.859375
15        235                                      J.D. Salinger    3.846939
16        469           Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
17        630  William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
18        106                                          Dan Brown    3.741259
19        372                                         Lois Lowry    3.738462
20        195            George Orwell/Boris Grabnar/Peter Škerl    3.729730
21        554                                    Stephenie Meyer    3.662500
22        311                                     John Steinbeck    3.643836

1. Джоан Роулинг - Британская писательница
2. Агата Кристи -  Британская писательница
3. Маркус Зузак - Австралийский писатель
4. Джон Рональд Руэл Толкин - Британский писатель
5. Роальд Даль/Квентин Блейк - Британский писатель/Британский карикатурист
6. Луиза Мэй Олкотт - Американская писательница
7. Рик Риордан - Американский писатель
8. Артур Голден - Американский писатель
9. Стивен Кинг - Американский писатель
10. Джон Гришэм - Американский писатель
11. Уильям Голдинг - Британский писатель
12. Николас Спаркс - Американский писатель
13. Джоди Пиколт - Американская писательница
14. Софи Кинселла - Британская писательница
15. Джеймс Паттерсон - Американский писатель
16. Джером Дэвид Сэлинджер - Американский писатель
17. Пауло Коэльо/Кларк Ричард Алан/Оздемир Инс - Бразильский писатель/Американский публицист/Турецкий поэт
18. Уильям Шекспир/Пол Верстайн/Барбара А. Моуэт - Британский поэт/Редактор/Редактор
19. Дэн Браун - Американский писатель
20. Лоис Лоури - Американская писательница
21. Джордж Оруэлл/Борис Грабнар/Питер Скерл - Британский писатель/Переводчик/Шведский сценарист
22. Стефани Майер - Американская писательница
23. Джон Стейнбек - Американский писатель

Таковых нашлось 23 автора. И Стивен Кинг занимает отнюдь не второе место после Роулинг.

Стоит отметить следующее наблюдение - за исключением Пауло Коэльо (Бразильский писатель), Оздемир Инс (Турецкий поэт) и Питер Скерл (Шведский сценарист), все остальные представленные авторы являются представителями англоязычных стран. 

Вспомним условие поставленной задачи:

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Ни в коей мере не умаляя заслуги приведённых авторов и, более того, признавая несомненную гениальность некоторых их них, отметим всё же, что отсутствие европейских, латиноамериканских, российских и прочих авторов, может создать однобокость, на которую российские читатели по подписке не применут обратить внимание. Согласно вышеупомянутому наблюдению, в российской читательской среде данная база данных может иметь не столь высокий успех, как в англоязычных странах, где она, вероятно, и была приобретена.

[К содержанию](#content)

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок
[](#sql7)
<a id='sql7'></a>

In [25]:
query_14 = '''

        SELECT
            AVG(sub_review_cnt.review_cnt) AS avg_review_cnt
        FROM
            (SELECT
                COUNT(text) AS review_cnt
            FROM 
                reviews
            WHERE username IN
                (SELECT username
                FROM 
                    ratings
                GROUP BY
                    username
                HAVING
                    COUNT(ratings) > 50)
            GROUP BY 
                    username
            ) AS sub_review_cnt  
            
        '''
        
avg_count_review = pd.io.sql.read_sql(query_14, con = engine)
avg_count_review

avg_review_cnt
0       24.333333

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет 24 обзора.

[К содержанию](#content)

# Выводы:
[](#sql8)
<a id='sql8'></a>

- При выведении первых строк таблиц, отметим, что названия столбцов адекватно отображают своё содержимое. Явных признаков некорректного заполнения не обнаружено;
- По данным таблицы books, после первого января 2000 года вышла 821 книга;

Для каждой книги посчитали количество обзоров и среднюю оценку


Топ-5 книг (на самом деле первые пять книг таблицы books) с наибольшими показателями средней оценки, и максимумом обзоров:
- A Dirty Job (Grim Reaper #1) - Кристофер Мур "Грязная работа"
- Moneyball: The Art of Winning an Unfair Game - Майкл Льюис "Денежный шар"
- School's Out—Forever (Maximum Ride #2) - Джеймс Паттерсон "Максимальная поездка: Школа закрыта Навсегда"
- Act of Treason (Mitch Rapp #9) - Винс Флинн "Акт государственной измены"
- A Fistful of Charms (The Hollows #4) - Ким Харрисон "Пригоршня прелестей"	

Топ-5 книг с максимумом обзоров:
- Twilight (Twilight #1) - Стефани Майер "Сумерки";
- Outlander (Outlander #1) - Гэблдон Диана "Чужестранка";
- Eat Pray Love - Элизабет Гилберт "Есть, молиться, любить";
- Harry Potter and the Chamber of Secrets - Джоан Роулинг "Гарри Поттер и Тайная комната";
- Harry Potter and the Prisoner of Azkaban - Джоан Роулинг "Гарри Поттер и узник Азкабана".

Обратим внимание на то, что каждое из произведений, попавших в список первых строк по максимальному количеству обзоров, обязательно имеет экранизацию.

Топ-5 книг с наибольшими показателями средней оценки:
- A Woman of Substance (Emma Harte Saga #1)	- Барбара Брэдфорд "Состоятельная женщина";
- Act of Treason (Mitch Rapp #9) - Винс Флинн "Акт государственной измены"
- A Dirty Job (Grim Reaper #1) - Кристофер Мур "Грязная работа"
- A Fistful of Charms (The Hollows #4) - Ким Харрисон "Пригоршня прелестей"
- Alas Babylon	- Пэт Франк "Увы, Вавилон"

Обратим внимание на пересечение произведений по критерию самого высокого показателя среднего рейтинга и объединённого критерия - с наибольшими показателями средней оценки, и максимумом обзоров:
- Act of Treason (Mitch Rapp #9) - Винс Флинн "Акт государственной измены"
- A Dirty Job (Grim Reaper #1) - Кристофер Мур "Грязная работа"
- A Fistful of Charms (The Hollows #4) - Ким Харрисон "Пригоршня прелестей"

В таблице publishers представлено максимальное количество изданий издательства Penguin Books - 42 произведения толще 50 страниц;

Автор с самой высокой средней оценкой книг (с 50 и более оценками) - Джоан Роулинг. При выводе фамилий всех авторов с высокой средней оценкой книг (с учётом 50 или более оценок) мы получили список из 23 авторов:

1. Джоан Роулинг - Британская писательница
2. Агата Кристи -  Британская писательница
3. Маркус Зузак - Австралийский писатель
4. Джон Рональд Руэл Толкин - Британский писатель
5. Роальд Даль/Квентин Блейк - Британский писатель/Британский карикатурист
6. Луиза Мэй Олкотт - Американская писательница
7. Рик Риордан - Американский писатель
8. Артур Голден - Американский писатель
9. Стивен Кинг - Американский писатель
10. Джон Гришэм - Американский писатель
11. Уильям Голдинг - Британский писатель
12. Николас Спаркс - Американский писатель
13. Джоди Пиколт - Американская писательница
14. Софи Кинселла - Британская писательница
15. Джеймс Паттерсон - Американский писатель
16. Джером Дэвид Сэлинджер - Американский писатель
17. Пауло Коэльо/Кларк Ричард Алан/Оздемир Инс - Бразильский писатель/Американский публицист/Турецкий поэт
18. Уильям Шекспир/Пол Верстайн/Барбара А. Моуэт - Британский поэт/Редактор/Редактор
19. Дэн Браун - Американский писатель
20. Лоис Лоури - Американская писательница
21. Джордж Оруэлл/Борис Грабнар/Питер Скерл - Британский писатель/Переводчик/Шведский сценарист
22. Стефани Майер - Американская писательница
23. Джон Стейнбек - Американский писатель

Стоит отметить следующее наблюдение - за исключением Пауло Коэльо (Бразильский писатель), Оздемир Инс (Турецкий поэт) и Питер Скерл (Шведский сценарист), все остальные представленные авторы являются представителями англоязычных стран.

Вспомним условие поставленной задачи:

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Ни в коей мере не умаляя заслуги приведённых авторов и, более того, признавая несомненную гениальность некоторых их них, отметим всё же, что отсутствие европейских, латиноамериканских, российских и прочих авторов, может создать однобокость, на которую российские читатели по подписке не применут обратить внимание. Согласно вышеупомянутому наблюдению, в российской читательской среде данная база данных может иметь не столь высокий успех, как в англоязычных странах, где она, вероятно, и была приобретена.

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет 24 обзора.